pkg/apiserver/slowquery/queries.go

```diff
@@ -0,0 +1,157 @@
+	GetCommitTSTime    float64 `gorm:"column:Get_commit_ts_time" json:"get_commit_ts_time"`
+	LocalLatchWaitTime float64 `gorm:"column:Local_latch_wait_time" json:"local_latch_wait_time"`
+	ResolveLockTime    float64 `gorm:"column:Resolve_lock_time" json:"resolve_lock_time"`
+	PrewriteTime       float64 `gorm:"column:Prewrite_time" json:"prewrite_time"`
+	CommitTime         float64 `gorm:"column:Commit_time" json:"commit_time"`
+	CommitBackoffTime  float64 `gorm:"column:Commit_backoff_time" json:"commit_backoff_time"`
+	CopProcAvg         float64 `gorm:"column:Cop_proc_avg" json:"cop_proc_avg"`
+	CopProcP90         float64 `gorm:"column:Cop_proc_p90" json:"cop_proc_p90"`
+	CopProcMax         float64 `gorm:"column:Cop_proc_max" json:"cop_proc_max"`
+	CopWaitAvg         float64 `gorm:"column:Cop_wait_avg" json:"cop_wait_avg"`
+	CopWaitP90         float64 `gorm:"column:Cop_wait_p90" json:"cop_wait_p90"`
+	CopWaitMax         float64 `gorm:"column:Cop_wait_max" json:"cop_wait_max"`
+
+	// Transaction
+	TxnStartTS     uint `gorm:"column:Txn_start_ts" json:"txn_start_ts"`
+	WriteKeys      int  `gorm:"column:Write_keys" json:"write_keys"`
+	WriteSize      int  `gorm:"column:Write_size" json:"write_size"`
+	PrewriteRegion int  `gorm:"column:Prewrite_region" json:"prewrite_region"`
+	TxnRetry       int  `gorm:"column:Txn_retry" json:"txn_retry"`
+
+	// Coprocessor
+	RequestCount uint   `gorm:"column:Request_count" json:"request_count"`
+	ProcessKeys  uint   `gorm:"column:Process_keys" json:"process_keys"`
+	TotalKeys    uint   `gorm:"column:Total_keys" json:"total_keys"`
+	CopProcAddr  string `gorm:"column:Cop_proc_addr" json:"cop_proc_addr"`
+	CopWaitAddr  string `gorm:"column:Cop_wait_addr" json:"cop_wait_addr"`
+}
+
+func (b *Base) AfterFind() (err error) {
+	if !b.Time.IsZero() {
+		b.Timestamp = float64(b.Time.UnixNano()) / 1e9
+	}
+	return
+}
+
+type QueryRequestParam struct {
+	LogStartTS int64  `json:"logStartTS" form:"logStartTS"`
+	LogEndTS   int64  `json:"logEndTS" form:"logEndTS"`
+	DB         string `json:"db" form:"db"`
+	Limit      int    `json:"limit" form:"limit"`
+	Text       string `json:"text" form:"text"`
+	OrderBy    string `json:"orderBy" form:"orderBy"`
+	DESC       bool   `json:"desc" form:"desc"`
+}
+
+func QuerySlowLogList(db *gorm.DB, params *QueryRequestParam) ([]Base, error) {
+	tx := db.Table(SlowQueryTable)
+	tx = tx.Where("time between from_unixtime(?) and from_unixtime(?)", params.LogStartTS, params.LogEndTS)
+	if params.Text != "" {
+		tx = tx.Where("txn_start_ts REGEXP ? OR digest REGEXP ? OR prev_stmt REGEXP ? OR query REGEXP ?",
+			params.Text,
+			params.Text,
+			params.Text,
+			params.Text,
+		)
+	}
+
+	if params.DB != "" {
+		dbs := strings.Split(params.DB, ",")
+		tx = tx.Where("DB IN (?)", dbs)
+	}
+
+	order := params.OrderBy
+	if params.DESC {
+		// FIXME
+		// but grom.Expr("? DESC", order) doesn't work
+		tx = tx.Order(fmt.Sprintf("%s desc", order))
```

Comment:
we need to verify order is one of the field. Otherwise there will be SQL injection problems. You can use reflection as in statement app.